### PR TITLE
fix(ls): some weird git issue

### PR DESCRIPTION
**Pull request by ajbt200128**
When the language server would scan a file before it was saved, it'd scan a temporary file. This means none of the git filtering stuff would apply, and so a ton of findings that weren't relevant would pop up. Also untracked files in a git repo would not be scanned. This fixes both.

## Test Plan
Manually tested it by opening a file that has issues that have been committed already, then modifying without saving

### CODE REVIEW

File: changelog.d/pdx-ls-git.fixed [hunk 1]
```diff
@@ -0,0 +1 @@
+The language server will no longer surface committed findings when a user types but does not save
```
Comment (aryx) [inline changelog.d/pdx-ls-git.fixed:1]: I dont understand this sentence. What "surface" mean? How findings can be committed? (code is committed), if a user types then nothing is committed yet.

File: libs/git_wrapper/Git_wrapper.ml [hunk 2]
```diff
@@ -298,27 +298,42 @@ let is_git_repo cwd =
...
       in
       Option.bind lines (fun l -> Some (range_of_git_diff l))
-  | Ok _ -> None
+  | Ok _, _ -> None
+  | Error (`Msg e), _ -> raise (Error e)
+
+let is_tracked_by_git file =
+  let cwd = Fpath.parent file in
+  let cmd =
+    Bos.Cmd.(v "git" % "-C" % !!cwd % "ls-files" % "--error-unmatch" % !!file)
```
Comment (aryx) [inline libs/git_wrapper/Git_wrapper.ml:332]: small comment on --error-unmatch? What does it do?

File: src/osemgrep/language_server/scan_helpers/Processed_run.ml [hunk 3]
```diff
@@ -15,28 +15,33 @@ module Out = Semgrep_output_v1_t
 (* Helpers *)
 (*************************************************************************)
 
+(** Checks if the range of [cli_match] overlaps with [dirty_lines] *)
 let match_in_dirty_lines (cli_match : Out.cli_match) dirty_lines =
   Array.exists
     (fun (start, end_) ->
-      start <= cli_match.start.line && cli_match.start.line <= end_)
+      start <= cli_match.end_.line && cli_match.start.line <= end_)
```
Comment (aryx) [inline src/osemgrep/language_server/scan_helpers/Processed_run.ml:22]: range is difficult!

**Review by aryx — APPROVED** [review]

File: libs/git_wrapper/Git_wrapper.ml (hunk 2 above)
Comment (ajbt200128) [reply libs/git_wrapper/Git_wrapper.ml:332]: Returns a non 0 exit code if a file is not tracked by git. This way further on in this function we don't try running git diff on untracked files, as this isn't allowed.

File: src/osemgrep/language_server/scan_helpers/Processed_run.ml (hunk 3 above)
Comment (ajbt200128) [reply src/osemgrep/language_server/scan_helpers/Processed_run.ml:22]: Yes :(

File: libs/git_wrapper/Git_wrapper.ml (hunk 2 above)
Comment (ajbt200128) [reply libs/git_wrapper/Git_wrapper.ml:332]: will add a comment

File: changelog.d/pdx-ls-git.fixed (hunk 1 above)
Comment (ajbt200128) [reply changelog.d/pdx-ls-git.fixed:1]: I.e. display them in whatever editor is running. In the extension a finding that is strictly part of lines that have been committed and unchanged, aren't surfaced, similar to how pre-commit doesn't surface findings in files that weren't modified by the user. The bug is that they would be surfaced when a user would type, but disappear when saved, when the expected behavior is that when typing they wouldn't surface at all.

**Pull request merged**